File: leetcode-js/347-top-k-frequent-elements.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number[]}
 */
var topKFrequent = function(nums, k) {
    const buckets = Array(nums.length+1);
    const frequencyMap = new Map();

    for (const n of nums) {
        const count = frequencyMap.get(n);
        frequencyMap.set(n, count ? count + 1 : 1);
    }

    for (const [key, frequency] of frequencyMap.entries()) {
        if (!buckets[frequency]) {
            buckets[frequency] = [key];
        } else {
            buckets[frequency].push(key);
        }
    }

    const result = [];

    for (let i = buckets.length - 1; i >= 0, result.length < k; i -= 1) {
        if (buckets[i]) {
            result.push(...buckets[i]);
        }
    }

    return result;
};

File: leetcode-java/347-top-k-frequent-elements.java
public class Solution {
    public List<Integer> topKFrequent(int[] nums, int k) {
        List<Integer>[] bucket = new List[nums.length + 1];
        Map<Integer, Integer> frequencyMap = new HashMap<Integer, Integer>();

        for (int n : nums) {
            frequencyMap.put(n, frequencyMap.getOrDefault(n, 0) + 1);
        }

        for (int key : frequencyMap.keySet()) {
            int frequency = frequencyMap.get(key);
            if (bucket[frequency] == null) {
                bucket[frequency] = new ArrayList<>();
            }
            bucket[frequency].add(key);
        }

        List<Integer> res = new ArrayList<>();

        for (int i = bucket.length - 1; i >= 0 && res.size() < k; i--) {
            if (bucket[i] != null) {
                res.addAll(bucket[i]);
            }
        }

        return res;
    }
}
